smk/binning_preparation.smk

```diff
@@ -293,25 +303,45 @@ checkpoint make_assembly_batches:
     resources:
         mem = 5
     params:
-        min_length = lambda wildcards: wildcards.min_length
+        min_length = lambda wildcards: wildcards.min_length,
+        batch_filesize = calculate_batch_filesize
     run:
         import datetime
+        import os
 
         def logme(stream, msg):
             print(datetime.datetime.now().strftime("%d/%m/%Y %H:%M:%S"), msg, file=stream)
 
         with open(str(log), 'w') as f:
-            logme(f, "INFO: assemblies={} batch_size={}".format(len(input.assemblies), batch_size))
-            for i in range(0, len(input.assemblies), batch_size):
-                batch = 1 + int(i/batch_size)
-                logme(f, "INFO: Making BATCH={}".format(batch))
-
-                batch_input  = input.assemblies[i:i+batch_size]
-                batch_output = output.location + f"/batch{batch}.fasta.gz"
-                logme(f, "INFO:   INPUT ={}".format(batch_input))
-                logme(f, "INFO:   OUTPUT={}".format(batch_output))
-
-                filter_fasta_list_by_length(batch_input, batch_output, params.min_length)
+            logme(f, "INFO: assemblies={} batch_size={}Mb".format(len(input.assemblies), batch_filesize))
+            start = 0
+            batch = 1
+            current_batchfilesize = 0
+            for i in range(0, len(input.assemblies)):
+                current_assemblysize = os.path.getsize(input.assemblies[i]) / (1024**2)
+
+                # if adding the current assembly made batch larger than accepted, then write out batch
+                if (i and ((current_batchfilesize + current_assemblysize) > batch_filesize)) or (i+1 == len(input.assemblies)):
+                    logme(f, "INFO: Making BATCH={} size={:.0f}Mb".format(batch, current_batchfilesize))
+
+                    # take the final assembly too
+                    end = i
+                    if (i+1 == len(input.assemblies)):
+                        end = i+1
+                    # from start to end
+                    batch_input  = input.assemblies[start:end]
+                    batch_output = output.location + f"/batch{batch}.fasta.gz"
+                    logme(f, "INFO:   INPUT ={}".format(batch_input))
+                    logme(f, "INFO:   OUTPUT={}".format(batch_output))
+
+                    # write out batch
+                    filter_fasta_list_by_length(batch_input, batch_output, params.min_length)
+
+                    # reset variables
+                    start = i
+                    batch += 1
+                    current_batchfilesize = 0
+                current_batchfilesize += current_assemblysize
             logme(f, "INFO: done")
```

Comment:
Here's a suggestion for getting the filesize after length-filtering. This also avoids including the last file unconditionally, because it could break the memory limit.

```
        import datetime
        import os

        def logme(stream, msg):
            print(datetime.datetime.now().strftime("%d/%m/%Y %H:%M:%S"), msg, file=stream)

        n_assemblies = len(input.assemblies)
        with open(str(log), 'w') as f:
            logme(f, "INFO: assemblies={} batch_size={}Mb".format(n_assemblies, params.batch_filesize))
            start = 0
            batch = 1
            current_batchfilesize = 0
            for i in range(0, n_assemblies+1):
                # write current assembly and get size
                if i < n_assemblies:
                    filtered_fasta = f"asm_{i}.fasta.gz"
                    filter_fasta_by_length(input.assemblies[i], filtered_fasta, params.min_length)
                    current_assemblysize = os.path.getsize(filtered_fasta) / (1024**2)

                # if adding the current assembly made batch larger than accepted, then write out batch
                if (i and ((current_batchfilesize + current_assemblysize) > params.batch_filesize)) or (i == n_assemblies):
                    logme(f, "INFO: Making BATCH={} size={:.0f}Mb".format(batch, current_batchfilesize))

                    # from start to end
                    end = i
                    batch_input  = input.assemblies[start:end]
                    batch_output = output.location + f"/batch{batch}.fasta.gz"
                    logme(f, "INFO:   INPUT ={}".format(batch_input))
                    logme(f, "INFO:   OUTPUT={}".format(batch_output))

                    # write out batch by mimicing cat of gzipped files
                    # also remove the individual files once they are 'cat'ed
                    outfile = f"batch{batch}.fasta.gz"
                    with open(outfile, "wb") as out:
                        for j in range(start, end):
                            infile = f"asm_{j}.fasta.gz"
                            with open(infile, "rb") as inobj:
                                shutil.copyfileobj(inobj, out)
                            os.remove(infile)

                    # copy output batch file to destination
                    shutil.copy2(outfile, batch_output)

                    # reset variables
                    start = i
                    batch += 1
                    current_batchfilesize = 0

                # otherwise just continue counting
                current_batchfilesize += current_assemblysize

            logme(f, "INFO: done")
```

After writing this, I am thinking it might be easier to count bases while writing in `filter_fasta_by_length` and directly use the non-gzipped sizes.